File: src/field-segmentation.ts
import { fetchWithRetry } from "./analyze-parcel.js";

const FIELD_SEGMENTATION_MODEL_URL = process.env.FIELD_SEGMENTATION_MODEL_URL ?? "http://localhost:8000";
const EXTERNAL_REQUEST_TIMEOUT_MS = 30_000;

export interface FieldSegmentationResult {
  threshold: number;
  fieldPercentage: number;
  maskShape: number[];
  maskPngBase64: string;
  probabilityPngBase64: string;
}


export async function callFieldSegmentationModel(
  fileBytes: Buffer,
  filename: string,
  threshold = 0.5,
): Promise<FieldSegmentationResult> {
  const formData = new FormData();

  formData.append("file", new Blob([new Uint8Array(fileBytes)]), filename);

  const url = `${FIELD_SEGMENTATION_MODEL_URL}/predict?threshold=${threshold}`;
  console.log("Calling field segmentation model /predict at:", url);

  const resp = await fetchWithRetry(url, { method: "POST", body: formData }, EXTERNAL_REQUEST_TIMEOUT_MS);

  if (!resp.ok) {
    const errText = await resp.text().catch(() => "");
    console.error("Field segmentation /predict error:", resp.status, errText.slice(0, 500));
    throw new Error(`Modèle de segmentation indisponible : HTTP ${resp.status}`);
  }

  const data: unknown = await resp.json();
  if (!data || typeof data !== "object") throw new Error("Réponse du modèle de segmentation invalide.");

  const values = data as Record<string, unknown>;
  if (
    typeof values.field_percentage !== "number"
    || !Number.isFinite(values.field_percentage)
    || !Array.isArray(values.mask_shape)
    || typeof values.mask_png_base64 !== "string"
    || typeof values.probability_png_base64 !== "string"
  ) {
    throw new Error("La réponse du modèle de segmentation ne contient pas les valeurs attendues.");
  }

  return {
    threshold: typeof values.threshold === "number" ? values.threshold : threshold,
    fieldPercentage: values.field_percentage,
    maskShape: values.mask_shape as number[],
    maskPngBase64: values.mask_png_base64,
    probabilityPngBase64: values.probability_png_base64,
  };
}

File: app/services/field_segmentation.py
"""Port 1:1 de src/field-segmentation.ts : client HTTP trivial vers le modèle
externe de segmentation de parcelles (U-Net, service FastAPI séparé)."""

from __future__ import annotations

import logging
import math
from typing import Any

from app.config import settings
from app.services.analyze_parcel import fetch_with_retry

logger = logging.getLogger("agrisat.field_segmentation")

FIELD_SEGMENTATION_MODEL_URL = settings.field_segmentation_model_url or "http://localhost:8000"
EXTERNAL_REQUEST_TIMEOUT_S = 30.0


async def call_field_segmentation_model(file_bytes: bytes, filename: str, threshold: float = 0.5) -> dict[str, Any]:
    url = f"{FIELD_SEGMENTATION_MODEL_URL}/predict?threshold={threshold}"
    logger.info("Calling field segmentation model /predict at: %s", url)

    resp = await fetch_with_retry(
        "POST", url, files={"file": (filename, file_bytes)}, timeout_s=EXTERNAL_REQUEST_TIMEOUT_S
    )

    if resp.status_code >= 400:
        err_text = resp.text[:500] if resp.text else ""
        logger.error("Field segmentation /predict error: %s %s", resp.status_code, err_text)
        raise RuntimeError(f"Modèle de segmentation indisponible : HTTP {resp.status_code}")

    data = resp.json()
    if not isinstance(data, dict):
        raise RuntimeError("Réponse du modèle de segmentation invalide.")

    field_percentage = data.get("field_percentage")
    mask_shape = data.get("mask_shape")
    mask_png_base64 = data.get("mask_png_base64")
    probability_png_base64 = data.get("probability_png_base64")
    if (
        not isinstance(field_percentage, (int, float))
        or isinstance(field_percentage, bool)
        or not math.isfinite(field_percentage)
        or not isinstance(mask_shape, list)
        or not isinstance(mask_png_base64, str)
        or not isinstance(probability_png_base64, str)
    ):
        raise RuntimeError("La réponse du modèle de segmentation ne contient pas les valeurs attendues.")

    response_threshold = data.get("threshold")

    return {
        "threshold": response_threshold if isinstance(response_threshold, (int, float)) and not isinstance(response_threshold, bool) else threshold,
        "fieldPercentage": field_percentage,
        "maskShape": mask_shape,
        "maskPngBase64": mask_png_base64,
        "probabilityPngBase64": probability_png_base64,
    }
